File: src/test/java/cube8540/oauth/authentication/resource/infra/SecuredResourceExceptionTranslatorTest.java
package cube8540.oauth.authentication.resource.infra;

import cube8540.oauth.authentication.resource.domain.ResourceInvalidException;
import cube8540.oauth.authentication.resource.domain.ResourceNotFoundException;
import cube8540.oauth.authentication.resource.domain.ResourceRegisterException;
import cube8540.oauth.authentication.error.message.ErrorCodes;
import cube8540.oauth.authentication.error.message.ErrorMessage;
import cube8540.validator.core.ValidationError;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

@DisplayName("자원 예외 변환기 클래스 테스트")
class SecuredResourceExceptionTranslatorTest {
    private static final String DESCRIPTION = "DESCRIPTION";

    private SecuredResourceExceptionTranslator translator;

    @BeforeEach
    void setup() {
        this.translator = new SecuredResourceExceptionTranslator();
    }

    @Test
    @DisplayName("ResourceNotFoundException 변환")
    void translateResourceNotFoundException() {
        ResourceNotFoundException e = ResourceNotFoundException.instance(DESCRIPTION);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertNotNull(response.getBody());
        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
        assertEquals(ErrorCodes.NOT_FOUND, response.getBody().getErrorCode());
        assertEquals(DESCRIPTION, response.getBody().getDescription());
    }

    @Test
    @DisplayName("ResourceRegisterException 변환")
    void translateResourceRegisterException() {
        ResourceRegisterException e = ResourceRegisterException.existsIdentifier(DESCRIPTION);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertNotNull(response.getBody());
        assertEquals(HttpStatus.BAD_REQUEST, response.getStatusCode());
        assertEquals(ErrorCodes.EXISTS_IDENTIFIER, response.getBody().getErrorCode());
        assertEquals(DESCRIPTION, response.getBody().getDescription());
    }

    @Test
    @DisplayName("ResourceInvalidException 변환")
    @SuppressWarnings("unchecked")
    void translateResourceInvalidException() {
        List<ValidationError> errors = Arrays.asList(new ValidationError("p", "v"), new ValidationError("p", "v1"), new ValidationError("p", "v2"));
        ResourceInvalidException e = ResourceInvalidException.instance(errors);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertNotNull(response.getBody());
        assertEquals(HttpStatus.BAD_REQUEST, response.getStatusCode());
        assertEquals(ErrorCodes.INVALID_REQUEST, response.getBody().getErrorCode());
        assertArrayEquals(errors.toArray(), ((Collection<ValidationError>) Objects.requireNonNull(response.getBody().getDescription())).toArray());
    }

    @Test
    @DisplayName("기타 예외 상황 일시")
    void translateException() {
        Exception exception = new Exception();

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(exception);
        assertNotNull(response.getBody());
        assertEquals(HttpStatus.INTERNAL_SERVER_ERROR, response.getStatusCode());
        assertEquals(ErrorCodes.SERVER_ERROR, response.getBody().getErrorCode());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/resource/infra/SecuredResourceExceptionTranslatorTest.kt
package cube8540.oauth.authentication.resource.infra

import cube8540.oauth.authentication.resource.domain.ResourceInvalidException
import cube8540.oauth.authentication.resource.domain.ResourceNotFoundException
import cube8540.oauth.authentication.resource.domain.ResourceRegisterException
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpStatus
import java.util.stream.Stream

class SecuredResourceExceptionTranslatorTest {

    private val translator = SecuredResourceExceptionTranslator()

    @ParameterizedTest
    @MethodSource(value = ["resourceNotFoundExceptions"])
    fun `translate resource not found exception`(exception: ResourceNotFoundException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }

    @ParameterizedTest
    @MethodSource(value = ["resourceRegisterExceptions"])
    fun `translate resource register exception`(exception: ResourceRegisterException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @ParameterizedTest
    @MethodSource(value = ["resourceInvalidExceptions"])
    fun `translate resource invalid exception`(exception: ResourceInvalidException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    fun resourceNotFoundExceptions() = Stream.of(
        Arguments.of(ResourceNotFoundException.instance("resource not found"))
    )

    fun resourceRegisterExceptions() = Stream.of(
        Arguments.of(ResourceRegisterException.existsIdentifier("exists identifier"))
    )

    fun resourceInvalidExceptions() = Stream.of(
        Arguments.of(ResourceInvalidException.instance(emptyList()))
    )

}